ReactiveCocoa/Swift/Flatten.swift

```diff
@@ -0,0 +1,571 @@
+				state.observer.sendFailed(error)
+
+			case .Completed:
+				state.decrementInFlight()
+
+			case .Interrupted:
+				state.observer.sendInterrupted()
+			}
+		}
+	}
+}
+
+extension SignalProducerType where Value: SignalProducerType, Error == Value.Error {
+	/// Merges a `signal` of SignalProducers down into a single signal, biased toward the producer
+	/// added earlier. Returns a Signal that will forward events from the inner producers as they arrive.
+	@warn_unused_result(message="Did you forget to call `start` on the producer?")
+	private func merge() -> SignalProducer<Value.Value, Error> {
+		return SignalProducer<Value.Value, Error> { relayObserver, disposable in
+			let state = MergeState(observer: relayObserver, disposable: disposable)
+
+			self.startWithSignal { signal, signalDisposable in
+				disposable.addDisposable(signalDisposable)
+
+				signal.observeMerge(state)
+			}
+
+		}
+	}
+}
+
+extension SignalType {
+	/// Merges the given signals into a single `Signal` that will emit all values
+	/// from each of them, and complete when all of them have completed.
+	public static func merge<S: SequenceType where S.Generator.Element == Signal<Value, Error>>(signals: S) -> Signal<Value, Error> {
+		let producer = SignalProducer<Signal<Value, Error>, Error>(values: signals)
+		var result: Signal<Value, Error>!
+
+		producer.startWithSignal { (signal, _) in
+			result = signal.flatten(.Merge)
+		}
+
+		return result
+	}
+}
+
+public final class MergeState<Value, Error: ErrorType> {
+	/// The observer of a started `concat` producer.
+	let observer: Observer<Value, Error>
+
+	/// The top level disposable of a started `concat` producer.
+	let disposable: CompositeDisposable?
+
+	/// The number of outstanding merged signals
+	let inFlight = Atomic(1)
+
+	init(observer: Observer<Value, Error>, disposable: CompositeDisposable?) {
+		self.observer = observer
+		self.disposable = disposable
+	}
+
+	func decrementInFlight() {
+		let orig = inFlight.modify { $0 - 1 }
+		if orig == 1 {
+			observer.sendCompleted()
+		}
+	}
+}
+
+extension SignalType where Value: SignalProducerType, Error == Value.Error {
+	/// Returns a signal that forwards values from the latest signal sent on
+	/// `signal`, ignoring values sent on previous inner signal.
+	///
+	/// An error sent on `signal` or the latest inner signal will be sent on the
+	/// returned signal.
+	///
+	/// The returned signal completes when `signal` and the latest inner
+	/// signal have both completed.
+	@warn_unused_result(message="Did you forget to call `observe` on the signal?")
+	private func switchToLatest() -> Signal<Value.Value, Error> {
+		return Signal<Value.Value, Error> { observer in
+			self.observeSwitchToLatest(observer, Atomic(LatestState<Value, Error>()), SerialDisposable())
+		}
+	}
+
+	private func observeSwitchToLatest(observer: Observer<Value.Value, Error>, _ state: Atomic<LatestState<Value, Error>>, _ latestInnerDisposable: SerialDisposable) -> Disposable? {
```

Comment:
AFAICT there's not a good reason to pass in `state`. It looks like it's always initialized with a `Atomic(LatestState<Value, Error>())`.